### ISSUE TITLE
RangeError: Maximum call stack size exceeded at _deconstructPacket in CI Pipeline

### ISSUE THREAD

**Issue by poponuts** [cypress-io/cypress#15101]
### Current behavior
Upgrading to cypress version 6.5.0 and running it on Bitbucket Pipeline CI via parallel tests, the error is encountered:
```
RangeError: Maximum call stack size exceeded
    at _deconstructPacket (/root/.cache/Cypress/6.5.0/Cypress/resources/app/packages/socket/node_modules/socket.io-parser/dist/binary.js:21:28)
    at _deconstructPacket (/root/.cache/Cypress/6.5.0/Cypress/resources/app/packages/socket/node_modules/socket.io-parser/dist/binary.js:40:32)
...
at _deconstructPacket (/root/.cache/Cypress/6.5.0/Cypress/resources/app/packages/socket/node_modules/socket.io-The Test Runner unexpectedly exited via a exit event with signal SIGILL
```

I've checked similar issues in the past and they have been fixed by succeeding Cypress versions. Could it be on the package.json and package-lock.json having the latest version Cypress `v6.5.0` while my docker image is still pointing to `cypress/base:14.15.4`

### Desired behavior
The Test runner should not exit in the CI pipeline and continue to generate test results.

### Test code to reproduce
`bitbucket-pipelines.yml`
```yaml
image: cypress/base:14.15.4
options:
  max-time: 20

tests: &tests
  name: Run tests
  caches:
    - node
    - cypress
  script:
    - set +e # continue executing next steps regardless of errors
    - cypress_version=$HOST npx cypress run --record --key $CYPRESS_DASHBOARD_KEY --parallel --ci-build-id $BITBUCKET_BUILD_NUMBER
  artifacts:
    - cypress/screenshots/**
    - cypress/videos/**
    - cypress/test-reports/mocha/*.json
    - cypress/test-reports/junit.xml

pipelines:
  default:
  - step:
      name: Prepare dependencies
      caches:
        - npm
        - cypress
        - node
      script:
        - npm ci
        - npx @bahmutov/print-env BITBUCKET # let's see the environment variables right away
  - parallel: # run N steps in parallel
    - step:
        <<: *tests
    - step:
        <<: *tests
    - step:
        <<: *tests
definitions:
  caches:
    npm: $HOME/.npm
    cypress: $HOME/.cache/Cypress
```

### Versions
Cypress version 6.5.0
Running locally on MacOS works fine.
Running on Bitbucket pipeline CI is when the issue is encountered.

**Comment by ronnyek** [comment]
We too just started seeing this in our CI/CD pipelines just last night. I'm not sure if we recently just upgraded the version of cypress we target, but definitely happening now.

**Comment by syedmuhdahmad** [comment]
Reverting the cypress to 6.4.0 has fixed this issue for some reason. Only one thing is common between @poponuts  and my code is the definition part

> definitions:
  caches:
    npm: $HOME/.npm
    cypress: $HOME/.cache/Cypress
    node: node_modules

**Comment by CypressHarry** [comment]
A user has reported that these errors no longer appear after upgrading to version 6.6.0. Can others here confirm that the issue is solved upon upgrade?

**Comment by dingraham-dh** [comment]
@CypressHarry 
I am still seeing it after upgrading to 6.6.0. 

`RangeError: Maximum call stack size exceeded
    at _deconstructPacket (/root/.cache/Cypress/6.6.0/Cypress/resources/app/packages/socket/node_modules/socket.io-parser/dist/binary.js:21:28)`

**Comment by TimVee** [comment]
I have the same problem, but locally on Windows. I'm on Cypress 6.6.0.

**Comment by TimVee** [comment]
I got this bug to occur consistently in open mode by making a request with a timeout that is shorter than the request takes. I made a small repo with a reproducible example: https://github.com/TimVee/deconstructPacked-bug.

**Comment by dingraham-dh** [comment]
This bug is preventing our team from upgrading to any Cypress version past 6.4.
Is there any traction on this? Thanks.

**Comment by poponuts** [comment]
Now, with **6.6.0**, even on my local machine, it is producing same error. With 6.5.0, it was only on our CI pipeline. Trying to isolate the problem, It seems to fail on a spec file that uses `Cypress.Promise` . Further error message:
```at tryCatcher (/Users/evan.rosal/Library/Caches/Cypress/6.6.0/Cypress.app/Contents/Resources/app/packages/server/node_modules/bluebird/js/release/util.js:16:23)
    at Promise._settlePromiseFromHandler (/Users/evan.rosal/Library/Caches/Cypress/6.6.0/Cypress.app/Contents/Resources/app/packages/server/node_modules/bluebird/js/release/promise.js:547:31)
    at Promise._settlePromise (/Users/evan.rosal/Library/Caches/Cypress/6.6.0/Cypress.app/Contents/Resources/app/packages/server/node_modules/bluebird/js/release/promise.js:604:18)
    at Promise._settlePromise0 (/Users/evan.rosal/Library/Caches/Cypress/6.6.0/Cypress.app/Contents/Resources/app/packages/server/node_modules/bluebird/js/release/promise.js:649:10)
    at Promise._settlePromises (/Users/evan.rosal/Library/Caches/Cypress/6.6.0/Cypress.app/Contents/Resources/app/packages/server/node_modules/bluebird/js/release/promise.js:725:18)
    at _drainQueueStep (/Users/evan.rosal/Library/Caches/Cypress/6.6.0/Cypress.app/Contents/Resources/app/packages/server/node_modules/bluebird/js/release/async.js:93:12)
    at _drainQueue (/Users/evan.rosal/Library/Caches/Cypress/6.6.0/Cypress.app/Contents/Resources/app/packages/server/node_modules/bluebird/js/release/async.js:86:9)
    at Async._drainQueues (/Users/evan.rosal/Library/Caches/Cypress/6.6.0/Cypress.app/Contents/Resources/app/packages/server/node_modules/bluebird/js/release/async.js:102:5)
    at Immediate.Async.drainQueues [as _onImmediate] (/Users/evan.rosal/Library/Caches/Cypress/6.6.0/Cypress.app/Contents/Resources/app/packages/server/node_modules/bluebird/js/release/async.js:15:14)
    at processImmediate (internal/timers.js:456:21)
```

**Comment by poponuts** [comment]
Hi @CypressHarry I hope this is being looked at by the team. @TimVee already provided a reproducible repo. 

By the way, I get same error on tests/specs that does not contain `Cypress.Promise` too. Will need to downgrade to 6.4.0 as a workaround.

**Comment by gweaths** [comment]
I'm having this same isue on 6.6.0 will this all be fixed in 6.7.2?

**Comment by hmonglee** [comment]
Same issue with 6.7.1

**Comment by RobbeCl** [comment]
Same in 6.8.0

**Comment by hmonglee** [comment]
downgrade to 6.4.0 and it's work.... please fix this

**Comment by creage** [comment]
@jennifer-shehane can we have your attention to this issue please? It is a blocking one, preventing from upgrading to the latest Cypress for many people here.

**Comment by t00** [comment]
@jennifer-shehane I created a new issue as instructed by support and included an app to reproduce the same issue: #15629

**Comment by shutchison-bai** [comment]
Also experiencing this.  Downgraded to 6.4.0 for now.

**Comment by jasonw-rr** [comment]
I start seeing this in the past 2 weeks, and downgraded to v6.4.0 worked for me

**Comment by poponuts** [comment]
This is still an issue. @bahmutov @jennifer-shehane any inputs on this? (someone provided a replicable repo above)
For additional error message (so Cypress team can isolate the possible problematic packages):
```
at Socket.packet (/Users/poponuts/Library/Caches/Cypress/6.8.0/Cypress.app/Contents/Resources/app/packages/socket/node_modules/socket.io/dist/socket.js:161:21)
    at /Users/poponuts/Library/Caches/Cypress/6.8.0/Cypress.app/Contents/Resources/app/packages/socket/node_modules/socket.io/dist/socket.js:270:18
    at /Users/poponuts/Library/Caches/Cypress/6.8.0/Cypress.app/Contents/Resources/app/packages/server/lib/socket-base.js:312:28
    at tryCatcher (/Users/poponuts/Library/Caches/Cypress/6.8.0/Cypress.app/Contents/Resources/app/packages/server/node_modules/bluebird/js/release/util.js:16:23)
    at Promise._settlePromiseFromHandler (/Users/poponuts/Library/Caches/Cypress/6.8.0/Cypress.app/Contents/Resources/app/packages/server/node_modules/bluebird/js/release/promise.js:547:31)
    at Promise._settlePromise (/Users/poponuts/Library/Caches/Cypress/6.8.0/Cypress.app/Contents/Resources/app/packages/server/node_modules/bluebird/js/release/promise.js:604:18)
    at Promise._settlePromise0 (/Users/poponuts/Library/Caches/Cypress/6.8.0/Cypress.app/Contents/Resources/app/packages/server/node_modules/bluebird/js/release/promise.js:649:10)
    at Promise._settlePromises (/Users/poponuts/Library/Caches/Cypress/6.8.0/Cypress.app/Contents/Resources/app/packages/server/node_modules/bluebird/js/release/promise.js:725:18)
    at _drainQueueStep (/Users/poponuts/Library/Caches/Cypress/6.8.0/Cypress.app/Contents/Resources/app/packages/server/node_modules/bluebird/js/release/async.js:93:12)
    at _drainQueue (/Users/poponuts/Library/Caches/Cypress/6.8.0/Cypress.app/Contents/Resources/app/packages/server/node_modules/bluebird/js/release/async.js:86:9)
    at Async._drainQueues (/Users/poponuts/Library/Caches/Cypress/6.8.0/Cypress.app/Contents/Resources/app/packages/server/node_modules/bluebird/js/release/async.js:102:5)
    at Immediate.Async.drainQueues [as _onImmediate] (/Users/poponuts/Library/Caches/Cypress/6.8.0/Cypress.app/Contents/Resources/app/packages/server/node_modules/bluebird/js/release/async.js:15:14)
    at processImmediate (internal/timers.js:456:21)
```

**Comment by lookininward** [comment]
Have also been experiencing this issue on test suite that runs to successful completion locally and most of the time in CI (Github Actions).

**Comment by lookininward** [comment]
Downgrading to `6.4.0` has also solved the issue for us.

**Comment by prashant-kajale** [comment]
same issue on 6.8

**Comment by poponuts** [comment]
The issue still persists on 7.0.0

**Comment by t00** [comment]
@jennifer-shehane will a hot-fix release be available soon?

**Comment by mysticdevx** [comment]
Same problem occurs with:
Cypress v 7.0.0
MacOS 11.2.3

`RangeError: Maximum call stack size exceeded
    at _deconstructPacket (/root/.cache/Cypress/7.0.0/Cypress/resources/app/packages/socket/node_modules/socket.io-parser/dist/binary.js:40:32)`

**Comment by kamiloski** [comment]
Get this on 7.0.1

**Comment by shutchison-bai** [comment]
Getting this on 7.1.0.  Would like to use the new features in 7.1.0, but can't until this issue is resolved.

**Comment by AdeZwart** [comment]
We are also seeing this issue running the tests on our Azure DevOps CI pipelines. And we've upgraded to 7.0.1 and still have the issue there:
`RangeError: Maximum call stack size exceeded
    at _deconstructPacket (E:\agent-5\_work\4\s\.cache\Cypress\7.0.1\Cypress\resources\app\packages\socket\node_modules\socket.io-parser\dist\binary.js:21:28)
[...]
at _deconstructPacket (E:\agent-5\_work\4\s\.cache\Cypress\7.0.1\Cypress\resources\app\packages\socket\node_modules\socket.io-parser\dist\binary.js:40:32)
    at _deconstructPacket (E:\agent-5\_work\4\s\.cache\Cypress\7.0.1\Cypress\resources\app\packages\socket\node_modules\socket.io-parser\dist\binary.js:32:26)
at Object.deconstructPacket (E:\agent-5\_work\4\s\.cache\Cypress\7.0.1\Cypress\resources\app\packages\socket\node_modules\socket.io-parser\dist\binary.js:16:17)
    at Encoder.encodeAsBinary (E:\agent-5\_work\4\s\.cache\Cypress\7.0.1\Cypress\resources\app\packages\socket\node_modules\socket.io-parser\dist\index.js:81:41)
    at Encoder.encode (E:\agent-5\_work\4\s\.cache\Cypress\7.0.1\Cypress\resources\app\packages\socket\node_modules\socket.io-parser\dist\index.js:43:29)
    at Client._packet (E:\agent-5\_work\4\s\.cache\Cypress\7.0.1\Cypress\resources\app\packages\socket\node_modules\socket.io\dist\client.js:167:44)
    at Socket.packet (E:\agent-5\_work\4\s\.cache\Cypress\7.0.1\Cypress\resources\app\packages\socket\node_modules\socket.io\dist\socket.js:161:21)
    at E:\agent-5\_work\4\s\.cache\Cypress\7.0.1\Cypress\resources\app\packages\socket\node_modules\socket.io\dist\socket.js:270:18
    at E:\agent-5\_work\4\s\.cache\Cypress\7.0.1\Cypress\resources\app\packages\server\lib\socket-base.js:325:28
    at tryCatcher (E:\agent-5\_work\4\s\.cache\Cypress\7.0.1\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\util.js:16:23)
    at Promise._settlePromiseFromHandler (E:\agent-5\_work\4\s\.cache\Cypress\7.0.1\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\promise.js:547:31)
    at Promise._settlePromise (E:\agent-5\_work\4\s\.cache\Cypress\7.0.1\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\promise.js:604:18)
    at Promise._settlePromise0 (E:\agent-5\_work\4\s\.cache\Cypress\7.0.1\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\promise.js:649:10)
    at Promise._settlePromises (E:\agent-5\_work\4\s\.cache\Cypress\7.0.1\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\promise.js:725:18)
    at _drainQueueStep (E:\agent-5\_work\4\s\.cache\Cypress\7.0.1\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\async.js:93:12)
    at _drainQueue (E:\agent-5\_work\4\s\.cache\Cypress\7.0.1\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\async.js:86:9)
    at Async._drainQueues (E:\agent-5\_work\4\s\.cache\Cypress\7.0.1\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\async.js:102:5)
    at Immediate.Async.drainQueues [as _onImmediate] (E:\agent-5\_work\4\s\.cache\Cypress\7.0.1\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\async.js:15:14)
    at processImmediate (internal/timers.js:461:21)`

It doesn't seem consistent in our case. Usually if I run the pipeline again it will be fine.

**Comment by bahmutov** [comment]
I do not see the issue happening locally on Mac using https://github.com/TimVee/deconstructPacked-bug repo (unfortunately).

**Comment by Prashika** [comment]
Hi, 
Deleting node-modules, package-lock.json and downgrading cypress to 6.4.0 resolved the issue for me, getting code-coverage now. But still cypress test runner crashes after running tests.

**Comment by poponuts** [comment]
I was on the verge of trying to isolate the issue in my case (down to `cy.request` auth value) but I just didn't have time to further investigate as there is an urgent need for us to run our regression suite, so I had no choice but to downgrade to **6.4.0**.

If I find time again in the future, I will look into it. However, if Cypress team (@bahmutov @jennifer-shehane @CypressHarry ) is able to further investigate into it, this is my line of code that is not working (then cross-check it against changes in 6.4.0+, maybe?):
```
Cypress.Commands.add('systems', (token) => { // token is grabbed from another cy.request custom command cy.auth
  return cy.request({
      method: 'POST',
      url: 'https://mysite.com/api/systems',
      auth: {
        bearer: token // <-- this is where I get that error (if I hard-code it to a value, it runs ok)
      },
      body: {
        'name': 'hello',
      },
      headers: {
        'Content-Type': 'application/json'
      },
      failOnStatusCode: false
  });
});

Cypress.Commands.add('auth', () => { // api for generating authentication tokens
  return cy.api({
      method: 'POST',
      url: 'https://mysite.com/api/oauth/token',
      body: {
        'client_id': clientId,
        'client_secret': clientSecret,
        'grant_type': grantType,
        'username': username,
        'password': password
      },
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded'
      }
  });
});
```

**Spec file:**
```
before(() => {
  cy.auth().should((respAuth) => {
    const token = respAuth.body.access_token;
    cy.systems(token);
  });
});
```

I know it is not a proper repository that can be replicated but I'll try my best to find time to create a boilerplate one.

**Comment by iondrimba** [comment]
Reproducible repo https://github.com/iondrimba/cypress_maximum_call_stack  Cypress 7.1.0

**Comment by t00** [comment]
@iondrimba Indeed crashes just fine on Arch Linux as well, I just needed to change git repo url to https://github.com/iondrimba/cypress_maximum_call_stack.git

**Comment by CinkadeusBG** [comment]
Same situation here.  Running 5.6 fine, but cannot update to 7.1.  Its killing our CI in DevOps.

**Comment by poponuts** [comment]
Has anyone tried this on 7.2.0? I have downgraded to 6.4.0 and its been smooth sailing so far. There were some performance issues fixed on 7.2.0 so it might have fixed it too.

**Comment by iondrimba** [comment]
@poponuts I've tried 7.2.0 on this repo https://github.com/iondrimba/cypress_maximum_call_stack  and the error still present.

**Comment by teebu** [comment]
Just started seeing this after upgrading to 7.2

**Comment by jennifer-shehane** [comment]
@iondrimba Thank you for a reproducible example. I have seen this error in our own Cypress tests sometimes, it has not been reliable however. This example is great because it will error each time. But it relies on GitHub's website which is not great because whatever is on github's site could change and the test may not fail in the future.

### Reproducible Example

`package.json`

```json
{
  "dependencies": {
    "@percy/cli": "*",
    "@percy/cypress": "^3.1.0",
    "cypress": "^7.1.0"
  }
}
```

```js
// support/index.js
import '@percy/cypress'
```

```js
// spec.js
it('not working with percy snapshot', () => {
  cy.visit('https://github.com/')
  cy.percySnapshot()
})
```

<img width="1429" alt="Screen Shot 2021-04-27 at 1 18 43 PM" src="https://user-images.githubusercontent.com/1271364/116292268-20422380-a75b-11eb-8f62-d01ee1f82e5c.png">
<img width="1593" alt="Screen Shot 2021-04-27 at 1 19 20 PM" src="https://user-images.githubusercontent.com/1271364/116292348-364fe400-a75b-11eb-8421-e88e7ad9bb77.png">

**Comment by iondrimba** [comment]
> @iondrimba Thank you for a reproducible example. I have seen this error in our own Cypress tests sometimes, it has not been reliable however. This example is great because it will error each time. But it relies on GitHub's website which is not great because whatever is on github's site could change and the test may not fail in the future.
> 
> ### Reproducible Example
> `package.json`
> 
> ```json
> {
>   "dependencies": {
>     "@percy/cli": "*",
>     "@percy/cypress": "^3.1.0",
>     "cypress": "^7.1.0"
>   }
> }
> ```
> 
> ```js
> // support/index.js
> import '@percy/cypress'
> ```
> 
> ```js
> // spec.js
> it('not working with percy snapshot', () => {
>   cy.visit('https://github.com/')
>   cy.percySnapshot()
> })
> ```
> 
> <img alt="Screen Shot 2021-04-27 at 1 18 43 PM" width="1429" src="https://user-images.githubusercontent.com/1271364/116292268-20422380-a75b-11eb-8f62-d01ee1f82e5c.png">
> 
> <img alt="Screen Shot 2021-04-27 at 1 19 20 PM" width="1593" src="https://user-images.githubusercontent.com/1271364/116292348-364fe400-a75b-11eb-8421-e88e7ad9bb77.png">

Indeed relying on github's page content is fragile, thanks for the advice @jennifer-shehane :+1:

**Comment by jennifer-shehane** [comment]
This is also a reproducible example although a bit more finicky to reproduce since you have to stop/start the test in open mode. https://github.com/cypress-io/cypress/issues/15629#issuecomment-823492885

**Comment by danielxdam** [comment]
Upgrading from 4.0.1 to 7.2 today. Stepped through each major version between and noticed no problems locally but observed this issue in the aws/codebuild/standard:5.0 that runs it in the CI/CD on 6.9.1 and 7.2.0. Switching to 6.4.0 seems to resolve it for now.

**Comment by jennifer-shehane** [comment]
This is a regression introduced in 6.5.0. 

[6.5.0 commits](https://github.com/cypress-io/cypress/compare/v6.4.0...v6.5.0). I've narrowed it down to this PR where it was introduced: https://github.com/cypress-io/cypress/pull/14479 which is unfortunate because it is a gigantic commit, but hopefully this will help us track it down.

**Issue not present:** https://github.com/cypress-io/cypress/commit/0abb5efe90b7c573df9d509b6ebe5d4debd3dab2
**Issue present:** https://github.com/cypress-io/cypress/commit/af26fbebe6bc609132013a0493a116cc78bb1bd4

**Comment by jennifer-shehane** [comment]
This may be due to the socket.io updates in the PR which required some patches to handled specific circular references.

**Comment by gweaths** [comment]
Is this something that can be reported and fixed ? It would be great to be able to get from 6.40 > latest version :-)
